### DISCUSSION TITLE
MCP230XX on ESP32

### DISCUSSION THREAD

**Discussion by Mark-Muc** [arendst/Tasmota#19218] [Q&A]
Hello .. I have  short question.
I've connected a MCP23017 to a W32 WROOM 32E board.
```
GPIO21 -> SDA
GPIO22 -> SCL
```
MCP23017 is recognized @ 0x20
```
10:23:47.634 CMD: I2CScan
10:23:47.663 MQT: stat/LCN_12_Port_M5/RESULT = {"I2CScan":"Device(s) found on bus1 at 0x20"}
```
nothing else is connected so far.
The problem is: **ALL Dxx are set default to "1"**
`10:25:32.995 MQT: stat/LCN_12_Port_M5/STATUS10 = {"StatusSNS":{"Time":"2023-07-30T10:25:32","MCP230XX":{"D0":1,"D1":1,"D2":1,"D3":1,"D4":1,"D5":1,"D6":1,"D7":1,"D8":1,"D9":1,"D10":1,"D11":1,"D12":1,"D13":1,"D14":1,"D15":1}}}`

if i connect a switch;  GND | D0 and press it D0 goes to "0"
```
10:27:19.792 CMD: sensor29 0,?
10:27:19.802 MQT: stat/LCN_12_Port_M5/RESULT = {"Sensor29_D0":{"MODE":2,"PULL_UP":"ON","INT_MODE":"ALL","STATE":"ON"}}
```
```
10:28:26.941 MQT: stat/LCN_12_Port_M5/RESULT = {"Time":"2023-07-30T10:28:26","MCP230XX_INT":{"D0":0,"MS":4821613}}
10:28:26.951 MQT: stat/LCN_12_Port_M5/RESULT = {"Event":"Done"}
```
release: D0 goes back to "1"
All hints say it it schoul toggle vrom "0" to "1".
Anybody can tell me whats wrong?
I am using this board:
![Screenshot 2023-07-27 083412](https://github.com/arendst/Tasmota/assets/67410847/1a2c501d-7a6f-4f59-aa55-066caac78d35)

Thanks for any help!!
Regards
Mark_Muc

**Comment by sfromis** [comment]
I'm not using any of those port extenders, but I observe that the docs talk about the option of configuring as inverted.
https://tasmota.github.io/docs/MCP230xx/

**Reply by Mark-Muc** [reply]
thanks .. I did read this doc several times .. but I am not able to finde any switch / command to invert the output
also tried to conmnect interrupt output to eather Vcc or GND .. nothing chanches..
Mark_Muc

**Reply by barbudor** [reply]
I assume that you are using the new 'mode 2' approach with a template as describe in that section https://tasmota.github.io/docs/MCP230xx/#mode-2
Mode 2 is now the default

In you define a pin in the template as a 'Button' you have the choice of straight or inverted
If you define a pin in the template as a 'Switch' then the standard switchmode command would apply

**Reply by Mark-Muc** [reply]
Thx barbudor .. I am using "mode 1" and "Sensor29" commands
but anyhow tomorrow I'll try to use "mode 2"
Regards
Mark_Muc

**Reply by Mark-Muc** [reply]
Hello .. so i tried to use "mode 2".
I compiled a version (dev) with `#define USE_MCP23XXX_DRV`
created a file `mcp23x.dat` with `{"NAME":"MCP23017 expander","BASE":0,"GPIO":[160,161,162,163,164,165,166,167,168,169,170,171,0,0,0,0]}`
plain Ascii.. 
i will add 12 switches!!
added it to the filesystem:
![image](https://github.com/arendst/Tasmota/assets/67410847/95885340-3b13-4e7f-b95f-67485761134f)
reboot .. now i can select in the GPIO conf. 
![image](https://github.com/arendst/Tasmota/assets/67410847/6c1a315f-6a52-41bb-9f6b-d4986797924b)
but i can only add 3 more GPIO .. (automatically added if selected) .. just working until GPIO39 .. no more and i only can select 6 interupts.
Can someone help me on this?
Thanks
Mark_Muc

**Reply by Mark-Muc** [reply]
so i solved it .. it was wrong in my thinking ...
i only have to select ONE (1) `mcp23xxx Int`  in the UI and THAN the  selection of the channels is in the file mcp23x.dat
Now it is working as I wished it to have!!
sorry for being so patient!!
Mark_Muc
PLS close!!

**Reply by barbudor** [reply]
Didn't do anything
You figured it out all by yourself 👍
